### ISSUE TITLE
To define a _DLL macro you need to include at least one system file

### ISSUE THREAD

**Issue by SudakovIvan** [mingw-w64/mingw-w64#53]
The code below does not compile because at least one system file must be included to define the **_DLL** macro:
```
//#include <iostream>
#ifdef _DLL
int main()
{
  std::cout << "yyyy" << std::endl;
  return 0;
}
#endif

```
As far as I know, the VS compiler defines the **_DLL** macro without the need to include system files

A real example can be seen here:

- https://github.com/apache/qpid-cpp/blob/main/src/qpid/client/windows/ClientDllMain.cpp
- https://github.com/apache/qpid-cpp/blob/main/src/qpid/sys/windows/QpidDllMain.h

**Comment by Biswa96** [comment]
You do not need to define or use `_DLL` macro to create DLL.

**Comment by SudakovIvan** [comment]
> You do not need to define or use `_DLL` macro to create DLL.

https://github.com/apache/qpid-cpp/blob/main/src/qpid/sys/windows/QpidDllMain.h - after mingw preprocessing, this file contains nothing because the **_DLL** is not defined, and that's the problem

**Comment by Biswa96** [comment]
> after mingw preprocessing, this file contains nothing because the **_DLL** is not defined, and that's the problem

That's the problem with the project - not with mingw. Try to define it manually or add that macro in the build system files in that project.

**Comment by SudakovIvan** [comment]
> > after mingw preprocessing, this file contains nothing because the **_DLL** is not defined, and that's the problem
> 
> That's the problem with the project - not with mingw. Try to define it manually or add that macro in the build system files in that project.

Ok, but with the VS compiler everything works in qpid...

**Comment by lhmouse** [comment]
`_DLL` is specific to MSVC and not defined by GCC: https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170

And for mingw-w64 there is no static CRT to link, so you can assume `_DLL` is always defined.

**Comment by SudakovIvan** [comment]
> `_DLL` is specific to MSVC and not defined by GCC: https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170
> 
> And for mingw-w64 there is no static CRT to link, so you can assume `_DLL` is always defined.

Yes, I agree. Just for the macro _DLL to be defined, you must do "#include" since this macro is defined explicitly here https://github.com/mingw-w64/mingw-w64/blob/48cf32beae4d400cf6a7f1d57ab51c2b409a9f44/mingw-w64-headers/crt/_mingw.h.in#L325, and in code without any "#include" the macro will not be defined, but in MSVC it will be.
Therefore, the behavior of the same code turns out differently - I didn't expect this...
